src/backend/distributed/master/master_citus_tools.c

```diff
@@ -0,0 +1,538 @@
+
+	for (index = 0; index < nodeNameCount; index++)
+	{
+		text *nodeNameText = DatumGetTextP(nodeNameDatumArray[index]);
+		char *nodeName = text_to_cstring(nodeNameText);
+		int32 nodePort = DatumGetInt32(nodePortDatumArray[index]);
+		text *commandText = DatumGetTextP(commandStringDatumArray[index]);
+		char *commandString = text_to_cstring(commandText);
+
+		nodeNames[index] = makeStringInfo();
+		commandStrings[index] = makeStringInfo();
+
+		appendStringInfo(nodeNames[index], "%s", nodeName);
+		nodePorts[index] = nodePort;
+		appendStringInfo(commandStrings[index], "%s", commandString);
+	}
+
+	*nodeNameArray = nodeNames;
+	*nodePortsArray = nodePorts;
+	*commandStringArray = commandStrings;
+	*parallel = parallelExecution;
+
+	return nodeNameCount;
+}
+
+
+/*
+ * ExecuteCommandsInParellelAndStoreResults connects to each node specified in
+ * nodeNameArray and nodePortArray, and executes command in commandStringArray
+ * in parallel fashion. Execution success status and result is reported for
+ * each command in statusArray and resultStringArray. Each array contains
+ * commandCount items.
+ */
+static void
+ExecuteCommandsInParallelAndStoreResults(StringInfo *nodeNameArray, int *nodePortArray,
+										 StringInfo *commandStringArray,
+										 bool *statusArray, StringInfo *resultStringArray,
+										 int commmandCount)
+{
+	int commandIndex = 0;
+	char *nodeUser = CurrentUserName();
+	PGconn **connectionArray = palloc0(commmandCount * sizeof(PGconn *));
+	int completedCount = 0;
+
+	/* establish connections */
+	for (commandIndex = 0; commandIndex < commmandCount; commandIndex++)
+	{
+		char *nodeName = nodeNameArray[commandIndex]->data;
+		int nodePort = nodePortArray[commandIndex];
+		PGconn *connection = ConnectToNode(nodeName, nodePort, nodeUser);
+		StringInfo queryResultString = resultStringArray[commandIndex];
+
+		statusArray[commandIndex] = true;
+
+		connectionArray[commandIndex] = connection;
+
+		if (connection == NULL)
+		{
+			appendStringInfo(queryResultString, "failed to connect to %s:%d", nodeName,
+							 (int) nodePort);
+			statusArray[commandIndex] = false;
+			completedCount++;
+		}
+	}
+
+	/* send queries at once */
+	for (commandIndex = 0; commandIndex < commmandCount; commandIndex++)
+	{
+		int querySent = 0;
+		PGconn *connection = connectionArray[commandIndex];
+		char *queryString = commandStringArray[commandIndex]->data;
+		StringInfo queryResultString = resultStringArray[commandIndex];
+
+		/*
+		 * If we don't have a connection, nothing to send, error string should already
+		 * been filled.
+		 */
+		if (connection == NULL)
+		{
+			continue;
+		}
+
+		querySent = PQsendQuery(connection, queryString);
+
+		if (querySent == 0)
+		{
+			StoreErrorMessage(connection, queryResultString);
+			statusArray[commandIndex] = false;
+			PQfinish(connection);
+			connectionArray[commandIndex] = NULL;
+			completedCount++;
+		}
+	}
+
+	/* check for query results */
```

Comment:
I think it'd have been perfectly fine to just call PQgetResult() (which blocks) on each connection here. After all this isn't useful for commands returning a lot of data (where waiting individually would be beneficial), and the commands will have been sent above.   But since you've already written the code...